src/protobuf-net.BuildTools/Internal/CodeGenSemanticModelParser.cs

```diff
@@ -2,170 +2,88 @@
+    private static void RecycleStack(Stack<(ISymbol compilerSymbol, object codeGenType)> value)
     {
         if (value is not null && s_SpareQueue is null)
         {
             value.Clear();
             s_SpareQueue = value;
         }
     }
+
     internal static void Parse(CodeGenSet set, ISymbol symbol)
     {
         CodeGenFile? file = null;
+        object? currentCodeGenType = null;
+        
         var pending = GetStack(); // prevent deep thread-stack dive by using an explicit stack
-        pending.Push(symbol);
+        pending.Push((compilerSymbol: symbol, codeGenType: null)!);
+        
         while (pending.Count > 0)
         {
-            var current = pending.Pop();
-            if (current is ITypeSymbol type)
+            var (currentSymbol, parentGenType) = pending.Pop();
+            if (currentSymbol is ITypeSymbol type)
             {
-                Debug.WriteLine(type.Name);
                 switch (type.TypeKind)
                 {
                     case TypeKind.Struct:
                     case TypeKind.Class:
-                        var attribs = type.GetAttributes();
-                        if (IsProtoContract(attribs))
+                        if (file is null)
                         {
-                            if (file is null)
-                            {
-                                var firstPath = symbol.DeclaringSyntaxReferences.FirstOrDefault()?.SyntaxTree?.FilePath;
-                                if (string.IsNullOrWhiteSpace(firstPath)) firstPath = "unknown";
+                            var firstPath = currentSymbol.DeclaringSyntaxReferences.FirstOrDefault()?.SyntaxTree?.FilePath;
+                            if (string.IsNullOrWhiteSpace(firstPath)) firstPath = "unknown";
 
-                                file = new CodeGenFile(firstPath!);
-                                set.Files.Add(file);
-                            }
-                            ParseMessage(set, file, type, attribs);
+                            file = new CodeGenFile(firstPath!);
+                            set.Files.Add(file);
                         }
+                        
+                        currentCodeGenType = ParseType(file, parentGenType, type);
                         break;
                 }
             }
 
-            if (current is INamespaceOrTypeSymbol container)
+            if (currentSymbol is IPropertySymbol property)
             {
-                foreach (var tm in container.GetMembers())
-                {
-                    switch (tm.Kind)
-                    {
-                        case SymbolKind.Namespace:
-                        case SymbolKind.NamedType:
-                            pending.Push(tm);
-                            break;
-                    }
-                    
-                }
+                currentCodeGenType = ParseProperty(parentGenType as CodeGenMessage, currentSymbol, property);
```

Comment:
it seems like this should be something that happens inside the `ParseType`, i.e. inside `ParseType` we step through the defined properties (and fields) ?

Reply:
After using stack approach, now we have a "class per symbol type" parsers. Again this comment and code is outdated